{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cb93dcf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dandreas/.conda/envs/esm3_peft/lib/python3.10/site-packages/esm/pretrained.py:113: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state_dict = torch.load(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ESMProtein(sequence='MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD', secondary_structure=None, sasa=None, function_annotations=None, coordinates=None, plddt=None, ptm=None, potential_sequence_of_concern=False)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dandreas/.conda/envs/esm3_peft/lib/python3.10/site-packages/esm/pretrained.py:46: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state_dict = torch.load(\n",
      "/home/dandreas/.conda/envs/esm3_peft/lib/python3.10/site-packages/esm/pretrained.py:57: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state_dict = torch.load(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--------------\n",
      "\n",
      "ESMProtein(sequence='MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD', secondary_structure='CCCCCCCCCCSCCCCCHHHGCGTCHCHGCCTESGCCCHHSCTTTCCHHHHHHHTHSCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCTCCCETCCCCECCETCCCECEEEEECSSSCCEEEEEEETTTTEEEESECEEECEEEEESCCCCTCCEEEEEEEECCGCCSHSCCEEETSHGHTCTCHSCECESCCEEEESCGEEEECETTSSCCTEECCCESCTCTTCCEEEEEEECHSCTSSTGGEGGCCEEEEEEEECTTSCEEEEECEEEEEESCTGCEHTBHHHHCHCCSCCCCCCCCTCCCCCCCCSCCCCCCCCCCTCTCCSHSCECCTSHHHHHIHHHHHHHCCHSHHHTCCTCCCCCCCCCECCCCCCCCCCCCCCTECCCCCCCC', sasa=[inf, inf, inf, 126.23686218261719, inf, 105.51325988769531, 131.81536865234375, 122.89801788330078, 97.90599822998047, 124.61527252197266, inf, 108.31156921386719, 119.83208465576172, inf, 75.34469604492188, inf, 94.59020233154297, 85.32980346679688, inf, 62.99298095703125, 87.79939270019531, 81.13764953613281, inf, inf, 83.6565933227539, 80.01443481445312, 96.84112548828125, inf, 108.62008666992188, 99.08616638183594, 111.37040710449219, inf, 80.697998046875, 115.313232421875, inf, 101.85989379882812, 91.95148468017578, inf, 69.81507110595703, inf, 116.44678497314453, 117.33126068115234, inf, inf, 78.14226531982422, 50.05149841308594, 98.99716186523438, 109.25852966308594, 71.2370376586914, 74.00745391845703, 111.97212219238281, 108.66533660888672, inf, inf, 102.36669158935547, inf, 126.28715515136719, 116.46282958984375, 55.38005447387695, 130.9521942138672, 120.51437377929688, inf, 83.57050323486328, 124.24544525146484, inf, inf, 121.4421157836914, inf, 98.21878814697266, 87.1572494506836, 115.75022888183594, 131.49264526367188, 121.14202117919922, 85.12503051757812, 130.79823303222656, 90.77333068847656, 132.57574462890625, 98.62583923339844, 88.47859191894531, 131.59555053710938, 126.8094253540039, 128.1788330078125, 98.55277252197266, 89.74583435058594, 131.25546264648438, 90.53941345214844, 129.3970947265625, 83.29444885253906, 106.87870788574219, 98.2783432006836, inf, 95.90147399902344, inf, 83.12471008300781, 87.47374725341797, 85.8785171508789, 85.00465393066406, 90.263916015625, 81.24920654296875, inf, inf, 70.7286605834961, inf, inf, 35.75712203979492, 64.24339294433594, 61.877685546875, 29.931838989257812, 9.655858993530273, 78.04032135009766, 6.41567850112915, 6.517711639404297, 15.154834747314453, 54.53285217285156, 60.48743438720703, 59.34798812866211, 44.92253112792969, 66.46176147460938, 41.86943054199219, inf, 51.18309783935547, 12.489826202392578, 55.20099639892578, 5.441422939300537, 42.827125549316406, 42.5034065246582, 17.793415069580078, 78.45923614501953, 75.45901489257812, 68.04642486572266, 74.06156921386719, 67.89570617675781, 24.194599151611328, 45.02101516723633, 12.059746742248535, 65.9359359741211, 30.08930206298828, 33.25939178466797, inf, 52.70214080810547, 9.46741771697998, 26.486268997192383, 2.9064154624938965, 30.74938201904297, 3.7455270290374756, 50.58636474609375, 9.113922119140625, 77.19688415527344, 57.74653244018555, 100.099609375, 14.063262939453125, 49.01951599121094, 115.4506607055664, 40.297306060791016, 6.097912311553955, 78.20718383789062, 2.703545093536377, 46.562705993652344, 2.313883066177368, 28.60206413269043, 3.159811496734619, 20.23537826538086, 18.510229110717773, 67.94602966308594, 64.46582794189453, 43.761474609375, inf, 68.46670532226562, inf, 61.751922607421875, 70.33880615234375, 44.88051986694336, 41.94159698486328, inf, inf, 17.248180389404297, 70.49179077148438, 66.69446563720703, inf, 77.58543395996094, inf, 27.37580108642578, 62.03338623046875, 71.19676971435547, 67.81436157226562, 76.36477661132812, 36.90262985229492, inf, 34.84613800048828, 50.96307373046875, 50.3297119140625, 71.05377197265625, 55.25897979736328, 29.36870002746582, 26.876728057861328, inf, 29.72212028503418, 68.11554718017578, 38.5390625, 64.16732788085938, 43.04296875, 73.65591430664062, 34.613914489746094, 50.467811584472656, 59.94424057006836, 32.70702362060547, 50.96717834472656, 83.57752227783203, inf, 68.5900650024414, 48.40107727050781, 33.27124786376953, inf, 47.457862854003906, 35.537078857421875, 24.418331146240234, 51.5511474609375, 18.754413604736328, 47.304046630859375, 42.120445251464844, 76.3458251953125, 59.32219696044922, 58.683494567871094, 78.07750701904297, 67.39933013916016, 48.214439392089844, 54.927940368652344, 68.14386749267578, 27.6227970123291, 29.058467864990234, 32.528419494628906, 13.341552734375, 35.42372512817383, 14.423985481262207, 33.120750427246094, 26.919816970825195, 36.494544982910156, 15.957794189453125, 57.84705352783203, 38.85127258300781, 61.63117599487305, 30.970012664794922, inf, 37.38109588623047, 32.282432556152344, inf, 53.460693359375, inf, 54.839447021484375, 31.786714553833008, 4.018892288208008, 31.562328338623047, 1.6173347234725952, 21.305131912231445, 2.128632068634033, 8.844877243041992, 1.443415641784668, 16.02952003479004, 27.547080993652344, 84.2964096069336, 92.16432189941406, 35.07245635986328, 80.75827026367188, 70.45298767089844, 33.5952033996582, 7.243764400482178, 74.68296813964844, 19.906286239624023, 30.843101501464844, 10.414167404174805, 49.20899963378906, 4.422121047973633, 79.95082092285156, 12.298702239990234, 22.485624313354492, 31.620899200439453, 24.815813064575195, 71.39215087890625, 42.34027099609375, inf, 63.46586990356445, inf, inf, 63.784027099609375, 61.92786407470703, 70.87155151367188, 78.87471008300781, 64.05976867675781, 74.87986755371094, inf, inf, inf, 59.8028678894043, 81.46705627441406, 98.74800872802734, inf, inf, inf, inf, 89.99837493896484, 128.07752990722656, 69.04236602783203, 73.80021667480469, 116.20172119140625, inf, inf, 80.58464813232422, inf, 107.57412719726562, 107.95269775390625, 98.17679595947266, 122.04103088378906, 104.56834411621094, 105.78277587890625, 93.97911071777344, 124.89547729492188, inf, 117.60606384277344, inf, inf, inf, 100.4796371459961, inf, 91.55681610107422, 47.19141387939453, 78.62047576904297, inf, inf, 70.97351837158203, 64.86187744140625, 77.12188720703125, 56.588592529296875, inf, 37.83733367919922, inf, 84.1082992553711, inf, 70.90451049804688, 76.80986785888672, 77.88134765625, 68.46097564697266, 89.70191192626953, 96.24398803710938, 75.84489440917969, 60.315711975097656, 87.07345581054688, 49.844764709472656, 78.58390808105469, 103.2099380493164, 106.4044418334961, 90.21163940429688, 79.94599914550781, 70.4442138671875, 88.84843444824219, 73.54115295410156, 65.92477416992188, inf, inf, 126.14826202392578, 72.44361114501953, 69.60635375976562, 97.49285125732422, inf, 85.46586608886719, inf, 87.67814636230469, 87.59524536132812, inf, inf, inf, 96.70613098144531, inf, inf, 69.89985656738281, inf, 98.28787231445312, 116.35746765136719, 89.5242919921875, inf, inf, inf, inf, inf, inf, inf, inf, 120.93578338623047, inf, 59.81907653808594, 132.99697875976562, 103.70968627929688, 97.45063018798828, inf], function_annotations=[FunctionAnnotation(label='binding transcription', start=115, end=277), FunctionAnnotation(label='dna templated', start=115, end=277), FunctionAnnotation(label='factor activity', start=115, end=277), FunctionAnnotation(label='of biosynthetic', start=115, end=277), FunctionAnnotation(label='of dna', start=115, end=277), FunctionAnnotation(label='of gene', start=115, end=277), FunctionAnnotation(label='of macromolecule', start=115, end=277), FunctionAnnotation(label='of metabolic', start=115, end=277), FunctionAnnotation(label='of nitrogen', start=115, end=277), FunctionAnnotation(label='of nucleobase', start=115, end=277), FunctionAnnotation(label='of primary', start=115, end=277), FunctionAnnotation(label='of rna', start=115, end=277), FunctionAnnotation(label='regulator activity', start=115, end=277), FunctionAnnotation(label='rna biosynthetic', start=115, end=277), FunctionAnnotation(label='templated', start=115, end=277), FunctionAnnotation(label='templated transcription', start=115, end=277), FunctionAnnotation(label='transcription factor', start=115, end=277), FunctionAnnotation(label='transcription regulator', start=115, end=277), FunctionAnnotation(label='amino', start=119, end=136), FunctionAnnotation(label='amino', start=152, end=159), FunctionAnnotation(label='amino', start=178, end=182), FunctionAnnotation(label='amino', start=212, end=222), FunctionAnnotation(label='amino', start=228, end=232), FunctionAnnotation(label='amino', start=264, end=273), FunctionAnnotation(label='amino acid', start=119, end=136), FunctionAnnotation(label='amino acid', start=152, end=159), FunctionAnnotation(label='amino acid', start=212, end=222), FunctionAnnotation(label='amino acid', start=228, end=232), FunctionAnnotation(label='amino acid', start=266, end=273), FunctionAnnotation(label='class', start=128, end=133), FunctionAnnotation(label='rna polymerase', start=119, end=273), FunctionAnnotation(label='factor 3', start=140, end=150), FunctionAnnotation(label='factor 3', start=163, end=200), FunctionAnnotation(label='factor 3', start=206, end=247), FunctionAnnotation(label='initiation', start=130, end=134), FunctionAnnotation(label='initiation', start=140, end=263), FunctionAnnotation(label='polymerase', start=120, end=275), FunctionAnnotation(label='specific', start=120, end=275), FunctionAnnotation(label='double', start=121, end=275), FunctionAnnotation(label='sequence specific', start=121, end=131), FunctionAnnotation(label='sequence specific', start=158, end=162), FunctionAnnotation(label='sequence specific', start=168, end=179), FunctionAnnotation(label='sequence specific', start=202, end=207), FunctionAnnotation(label='sequence specific', start=249, end=265), FunctionAnnotation(label='sequence specific', start=270, end=275), FunctionAnnotation(label='specific dna', start=121, end=147), FunctionAnnotation(label='specific dna', start=152, end=275), FunctionAnnotation(label='oxoacid', start=122, end=127), FunctionAnnotation(label='oxoacid', start=178, end=182), FunctionAnnotation(label='oxoacid', start=212, end=222), FunctionAnnotation(label='oxoacid', start=228, end=232), FunctionAnnotation(label='oxoacid', start=264, end=273), FunctionAnnotation(label='oxoacid metabolic', start=122, end=127), FunctionAnnotation(label='oxoacid metabolic', start=178, end=182), FunctionAnnotation(label='oxoacid metabolic', start=212, end=222), FunctionAnnotation(label='oxoacid metabolic', start=228, end=232), FunctionAnnotation(label='oxoacid metabolic', start=264, end=273), FunctionAnnotation(label='protein binding', start=155, end=162), FunctionAnnotation(label='sequence', start=261, end=265), FunctionAnnotation(label='site', start=264, end=271), FunctionAnnotation(label='transcription initiation', start=130, end=134), FunctionAnnotation(label='transcription initiation', start=141, end=147), FunctionAnnotation(label='transcription initiation', start=153, end=263), FunctionAnnotation(label='trna', start=261, end=265), FunctionAnnotation(label='methionine', start=141, end=147), FunctionAnnotation(label='methionine', start=157, end=163), FunctionAnnotation(label='methionine', start=168, end=247), FunctionAnnotation(label='methionine', start=263, end=271), FunctionAnnotation(label='sigma', start=141, end=147), FunctionAnnotation(label='sigma', start=153, end=157), FunctionAnnotation(label='sigma', start=163, end=201), FunctionAnnotation(label='sigma', start=206, end=263), FunctionAnnotation(label='region', start=236, end=242), FunctionAnnotation(label='region', start=248, end=262), FunctionAnnotation(label='stranded', start=236, end=242), FunctionAnnotation(label='stranded', start=248, end=262), FunctionAnnotation(label='stranded dna', start=236, end=242), FunctionAnnotation(label='stranded dna', start=248, end=262), FunctionAnnotation(label='synthetase', start=168, end=182), FunctionAnnotation(label='synthetase', start=202, end=207), FunctionAnnotation(label='synthetase', start=264, end=271), FunctionAnnotation(label='hydrolysis', start=168, end=179), FunctionAnnotation(label='carboxylic', start=264, end=271), FunctionAnnotation(label='6-phosphogluconate dehydrogenase, domain 2 (IPR013328)', start=144, end=148), FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=122, end=127), FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=141, end=164), FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=169, end=257), FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=263, end=273), FunctionAnnotation(label='Class II aldolase/adducin N-terminal (IPR001303)', start=141, end=147), FunctionAnnotation(label='Class II aldolase/adducin N-terminal (IPR001303)', start=169, end=201), FunctionAnnotation(label='Class II aldolase/adducin N-terminal (IPR001303)', start=206, end=247), FunctionAnnotation(label='Small GTPase (IPR001806)', start=264, end=271), FunctionAnnotation(label='T-box transcription factor (IPR001699)', start=168, end=179), FunctionAnnotation(label='T-box transcription factor (IPR001699)', start=264, end=271), FunctionAnnotation(label='S-adenosylmethionine synthetase superfamily (IPR022636)', start=168, end=179), FunctionAnnotation(label='T-box superfamily (IPR036960)', start=264, end=271)], coordinates=tensor([[[-42.1101,  56.5402,  61.6483],\n",
      "         [-42.6844,  57.5141,  62.5706],\n",
      "         [-44.2071,  57.4300,  62.5804],\n",
      "         ...,\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf]],\n",
      "\n",
      "        [[-42.7904,  57.9456,  62.2176],\n",
      "         [-43.8425,  56.9642,  61.9746],\n",
      "         [-45.2186,  57.5474,  62.2784],\n",
      "         ...,\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf]],\n",
      "\n",
      "        [[-45.7275,  58.1059,  62.4697],\n",
      "         [-46.8133,  57.5644,  63.2801],\n",
      "         [-48.0543,  58.4469,  63.1962],\n",
      "         ...,\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf]],\n",
      "\n",
      "        ...,\n",
      "\n",
      "        [[-42.0301,  31.7129,  27.1149],\n",
      "         [-40.7981,  31.9729,  27.8522],\n",
      "         [-41.0301,  31.8876,  29.3571],\n",
      "         ...,\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf]],\n",
      "\n",
      "        [[-41.7940,  33.4095,  28.1477],\n",
      "         [-41.3808,  33.0069,  29.4880],\n",
      "         [-41.7367,  34.0751,  30.5167],\n",
      "         ...,\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf]],\n",
      "\n",
      "        [[-40.2313,  32.6091,  29.6002],\n",
      "         [-40.1701,  33.3620,  28.3518],\n",
      "         [-38.9126,  34.2227,  28.2905],\n",
      "         ...,\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf],\n",
      "         [     inf,      inf,      inf]]]), plddt=tensor([0.4513, 0.4346, 0.4399, 0.4471, 0.4237, 0.4380, 0.4155, 0.4269, 0.4261,\n",
      "        0.3927, 0.4429, 0.4268, 0.4603, 0.4820, 0.4818, 0.4292, 0.4721, 0.5007,\n",
      "        0.4806, 0.4933, 0.5313, 0.5363, 0.5208, 0.5371, 0.5818, 0.5992, 0.5765,\n",
      "        0.5908, 0.6033, 0.6223, 0.6341, 0.6271, 0.6138, 0.6001, 0.5744, 0.5917,\n",
      "        0.5920, 0.5965, 0.5836, 0.5532, 0.5289, 0.4885, 0.4579, 0.5553, 0.5288,\n",
      "        0.5895, 0.5812, 0.6194, 0.5955, 0.5325, 0.6073, 0.6419, 0.5711, 0.5450,\n",
      "        0.6110, 0.6499, 0.6011, 0.6013, 0.5772, 0.5458, 0.5114, 0.4633, 0.4228,\n",
      "        0.4205, 0.4431, 0.4582, 0.4749, 0.4805, 0.5191, 0.5516, 0.5467, 0.5949,\n",
      "        0.6040, 0.6599, 0.7017, 0.7206, 0.7291, 0.7138, 0.7101, 0.7102, 0.6976,\n",
      "        0.7227, 0.7024, 0.7350, 0.7166, 0.6777, 0.6356, 0.5859, 0.5246, 0.4666,\n",
      "        0.4969, 0.4877, 0.4721, 0.4803, 0.4970, 0.4151, 0.4071, 0.4257, 0.3793,\n",
      "        0.3550, 0.3353, 0.3313, 0.3028, 0.2909, 0.2696, 0.2618, 0.2422, 0.2251,\n",
      "        0.2282, 0.2524, 0.2528, 0.2546, 0.2388, 0.2567, 0.2406, 0.2234, 0.2323,\n",
      "        0.2283, 0.2253, 0.2646, 0.2781, 0.2832, 0.3059, 0.2979, 0.3063, 0.3156,\n",
      "        0.3188, 0.3302, 0.3200, 0.3129, 0.3232, 0.3034, 0.3169, 0.3199, 0.3142,\n",
      "        0.3220, 0.3277, 0.3371, 0.3575, 0.3586, 0.3588, 0.3585, 0.3665, 0.3681,\n",
      "        0.3588, 0.3680, 0.3807, 0.3740, 0.4079, 0.4474, 0.4511, 0.4854, 0.5075,\n",
      "        0.4882, 0.4725, 0.4699, 0.4529, 0.4206, 0.4310, 0.3873, 0.3891, 0.3512,\n",
      "        0.3383, 0.3060, 0.2673, 0.2387, 0.2294, 0.2073, 0.2062, 0.1925, 0.2195,\n",
      "        0.2635, 0.2596, 0.2269, 0.2412, 0.2577, 0.2212, 0.2104, 0.2328, 0.2272,\n",
      "        0.2186, 0.2264, 0.2091, 0.2088, 0.2347, 0.1907, 0.1859, 0.2237, 0.2096,\n",
      "        0.2463, 0.2637, 0.2714, 0.2498, 0.2450, 0.2876, 0.2853, 0.2791, 0.2375,\n",
      "        0.2705, 0.2906, 0.3665, 0.3304, 0.3712, 0.3436, 0.3710, 0.3639, 0.3439,\n",
      "        0.3108, 0.3005, 0.2661, 0.3024, 0.2860, 0.3305, 0.3096, 0.3114, 0.3376,\n",
      "        0.3426, 0.3211, 0.3226, 0.2870, 0.3071, 0.3072, 0.3103, 0.3167, 0.2830,\n",
      "        0.2507, 0.2873, 0.3067, 0.3162, 0.3292, 0.3477, 0.3750, 0.3861, 0.3851,\n",
      "        0.4057, 0.3992, 0.4354, 0.4168, 0.3963, 0.4075, 0.4133, 0.4357, 0.4288,\n",
      "        0.4211, 0.4146, 0.4218, 0.3939, 0.4042, 0.4231, 0.4373, 0.4648, 0.4409,\n",
      "        0.4575, 0.4580, 0.4496, 0.4698, 0.4670, 0.4943, 0.4838, 0.4279, 0.4833,\n",
      "        0.5022, 0.5063, 0.4653, 0.4281, 0.4292, 0.4509, 0.4305, 0.4257, 0.4215,\n",
      "        0.4040, 0.4060, 0.3906, 0.3900, 0.3742, 0.3673, 0.3665, 0.3844, 0.3707,\n",
      "        0.3792, 0.3931, 0.3739, 0.3771, 0.3800, 0.4471, 0.4005, 0.3626, 0.3984,\n",
      "        0.4039, 0.3745, 0.3592, 0.3397, 0.3103, 0.3556, 0.3716, 0.4101, 0.4182,\n",
      "        0.4145, 0.5145, 0.5605, 0.5496, 0.5653, 0.5659, 0.6038, 0.6126, 0.6420,\n",
      "        0.6637, 0.6101, 0.6519, 0.6309, 0.5481, 0.5955, 0.6044, 0.6294, 0.6022,\n",
      "        0.5698, 0.5703, 0.5950, 0.5685, 0.5165, 0.5424, 0.4897, 0.4378, 0.4135,\n",
      "        0.3741, 0.3894, 0.3612, 0.3766, 0.3886, 0.3637, 0.3618, 0.3405, 0.3338,\n",
      "        0.3051, 0.3163, 0.3206, 0.3400, 0.3245, 0.3559, 0.4031, 0.3719, 0.3730,\n",
      "        0.4222, 0.4219, 0.4051, 0.4589, 0.4636, 0.4415, 0.4358, 0.5059, 0.4918,\n",
      "        0.4551, 0.4617, 0.4888, 0.4725, 0.4006, 0.4137, 0.4043, 0.4138, 0.3871,\n",
      "        0.3845, 0.3195, 0.3729, 0.3902, 0.4830, 0.4996, 0.5486, 0.5813, 0.5981,\n",
      "        0.5805, 0.5861, 0.5105, 0.5548, 0.5525, 0.5268, 0.5508, 0.5504, 0.5476,\n",
      "        0.6106, 0.6318, 0.6520, 0.6581, 0.6323, 0.5828, 0.5882, 0.4432, 0.4339,\n",
      "        0.3591, 0.3400, 0.2800, 0.2477, 0.2088, 0.2177]), ptm=tensor([0.2296], device='cuda:2'), potential_sequence_of_concern=False)\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "from esm.models.esm3 import ESM3\n",
    "from esm.sdk.api import (\n",
    "    ESMProtein,\n",
    "    SamplingConfig,\n",
    "    SamplingTrackConfig,\n",
    ")\n",
    "from esm.utils.constants.models import ESM3_OPEN_SMALL\n",
    "from esm.utils.structure.protein_chain import ProteinChain\n",
    "\n",
    "# Initialize the client\n",
    "client = ESM3.from_pretrained(ESM3_OPEN_SMALL, device=torch.device(\"cuda:2\"))\n",
    "\n",
    "# Load the protein\n",
    "protein = ESMProtein(sequence = 'MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD')\n",
    "\n",
    "print(protein)\n",
    "\n",
    "# Predict function\n",
    "protein_tensor = client.encode(protein)\n",
    "inference_output = client.forward_and_sample(\n",
    "    protein_tensor,\n",
    "    SamplingConfig(\n",
    "        sequence=SamplingTrackConfig(),\n",
    "        structure=SamplingTrackConfig(),\n",
    "        secondary_structure=SamplingTrackConfig(),\n",
    "        sasa=SamplingTrackConfig(),\n",
    "        function=SamplingTrackConfig(only_sample_masked_tokens=False),\n",
    "        \n",
    "    ),\n",
    ")\n",
    "protein_tensor_with_function = inference_output.protein_tensor\n",
    "protein_with_function = client.decode(protein_tensor_with_function)\n",
    "print('\\n--------------\\n')\n",
    "print(protein_with_function)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f195c8e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def split_annotations(annotations, residue_vocab):\n",
    "    interpro, keywords, residue = [], [], []\n",
    "    ipr_regex = re.compile(r\"IPR\\d+\")\n",
    "    for a in annotations:\n",
    "        if ipr_regex.search(a.label):\n",
    "            interpro.append(a)\n",
    "        elif a.label in residue_vocab:\n",
    "            residue.append(a)\n",
    "        else:\n",
    "            keywords.append(a)\n",
    "    return interpro, keywords, residue\n",
    "\n",
    "residue_vocab = client.tokenizers.residue_annotations._labels\n",
    "interpro_list, keyword_list, residue_list = split_annotations(\n",
    "    protein_with_function.function_annotations, residue_vocab\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9eae6c57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[FunctionAnnotation(label='6-phosphogluconate dehydrogenase, domain 2 (IPR013328)', start=144, end=148),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=122, end=127),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=141, end=164),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=169, end=257),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal domain superfamily (IPR036409)', start=263, end=273),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal (IPR001303)', start=141, end=147),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal (IPR001303)', start=169, end=201),\n",
       " FunctionAnnotation(label='Class II aldolase/adducin N-terminal (IPR001303)', start=206, end=247),\n",
       " FunctionAnnotation(label='Small GTPase (IPR001806)', start=264, end=271),\n",
       " FunctionAnnotation(label='T-box transcription factor (IPR001699)', start=168, end=179),\n",
       " FunctionAnnotation(label='T-box transcription factor (IPR001699)', start=264, end=271),\n",
       " FunctionAnnotation(label='S-adenosylmethionine synthetase superfamily (IPR022636)', start=168, end=179),\n",
       " FunctionAnnotation(label='T-box superfamily (IPR036960)', start=264, end=271)]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpro_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "870f43b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[FunctionAnnotation(label='binding transcription', start=115, end=277),\n",
       " FunctionAnnotation(label='dna templated', start=115, end=277),\n",
       " FunctionAnnotation(label='factor activity', start=115, end=277),\n",
       " FunctionAnnotation(label='of biosynthetic', start=115, end=277),\n",
       " FunctionAnnotation(label='of dna', start=115, end=277),\n",
       " FunctionAnnotation(label='of gene', start=115, end=277),\n",
       " FunctionAnnotation(label='of macromolecule', start=115, end=277),\n",
       " FunctionAnnotation(label='of metabolic', start=115, end=277),\n",
       " FunctionAnnotation(label='of nitrogen', start=115, end=277),\n",
       " FunctionAnnotation(label='of nucleobase', start=115, end=277),\n",
       " FunctionAnnotation(label='of primary', start=115, end=277),\n",
       " FunctionAnnotation(label='of rna', start=115, end=277),\n",
       " FunctionAnnotation(label='regulator activity', start=115, end=277),\n",
       " FunctionAnnotation(label='rna biosynthetic', start=115, end=277),\n",
       " FunctionAnnotation(label='templated', start=115, end=277),\n",
       " FunctionAnnotation(label='templated transcription', start=115, end=277),\n",
       " FunctionAnnotation(label='transcription factor', start=115, end=277),\n",
       " FunctionAnnotation(label='transcription regulator', start=115, end=277),\n",
       " FunctionAnnotation(label='amino', start=119, end=136),\n",
       " FunctionAnnotation(label='amino', start=152, end=159),\n",
       " FunctionAnnotation(label='amino', start=178, end=182),\n",
       " FunctionAnnotation(label='amino', start=212, end=222),\n",
       " FunctionAnnotation(label='amino', start=228, end=232),\n",
       " FunctionAnnotation(label='amino', start=264, end=273),\n",
       " FunctionAnnotation(label='amino acid', start=119, end=136),\n",
       " FunctionAnnotation(label='amino acid', start=152, end=159),\n",
       " FunctionAnnotation(label='amino acid', start=212, end=222),\n",
       " FunctionAnnotation(label='amino acid', start=228, end=232),\n",
       " FunctionAnnotation(label='amino acid', start=266, end=273),\n",
       " FunctionAnnotation(label='class', start=128, end=133),\n",
       " FunctionAnnotation(label='rna polymerase', start=119, end=273),\n",
       " FunctionAnnotation(label='factor 3', start=140, end=150),\n",
       " FunctionAnnotation(label='factor 3', start=163, end=200),\n",
       " FunctionAnnotation(label='factor 3', start=206, end=247),\n",
       " FunctionAnnotation(label='initiation', start=130, end=134),\n",
       " FunctionAnnotation(label='initiation', start=140, end=263),\n",
       " FunctionAnnotation(label='polymerase', start=120, end=275),\n",
       " FunctionAnnotation(label='specific', start=120, end=275),\n",
       " FunctionAnnotation(label='double', start=121, end=275),\n",
       " FunctionAnnotation(label='sequence specific', start=121, end=131),\n",
       " FunctionAnnotation(label='sequence specific', start=158, end=162),\n",
       " FunctionAnnotation(label='sequence specific', start=168, end=179),\n",
       " FunctionAnnotation(label='sequence specific', start=202, end=207),\n",
       " FunctionAnnotation(label='sequence specific', start=249, end=265),\n",
       " FunctionAnnotation(label='sequence specific', start=270, end=275),\n",
       " FunctionAnnotation(label='specific dna', start=121, end=147),\n",
       " FunctionAnnotation(label='specific dna', start=152, end=275),\n",
       " FunctionAnnotation(label='oxoacid', start=122, end=127),\n",
       " FunctionAnnotation(label='oxoacid', start=178, end=182),\n",
       " FunctionAnnotation(label='oxoacid', start=212, end=222),\n",
       " FunctionAnnotation(label='oxoacid', start=228, end=232),\n",
       " FunctionAnnotation(label='oxoacid', start=264, end=273),\n",
       " FunctionAnnotation(label='oxoacid metabolic', start=122, end=127),\n",
       " FunctionAnnotation(label='oxoacid metabolic', start=178, end=182),\n",
       " FunctionAnnotation(label='oxoacid metabolic', start=212, end=222),\n",
       " FunctionAnnotation(label='oxoacid metabolic', start=228, end=232),\n",
       " FunctionAnnotation(label='oxoacid metabolic', start=264, end=273),\n",
       " FunctionAnnotation(label='protein binding', start=155, end=162),\n",
       " FunctionAnnotation(label='sequence', start=261, end=265),\n",
       " FunctionAnnotation(label='site', start=264, end=271),\n",
       " FunctionAnnotation(label='transcription initiation', start=130, end=134),\n",
       " FunctionAnnotation(label='transcription initiation', start=141, end=147),\n",
       " FunctionAnnotation(label='transcription initiation', start=153, end=263),\n",
       " FunctionAnnotation(label='methionine', start=141, end=147),\n",
       " FunctionAnnotation(label='methionine', start=157, end=163),\n",
       " FunctionAnnotation(label='methionine', start=168, end=247),\n",
       " FunctionAnnotation(label='methionine', start=263, end=271),\n",
       " FunctionAnnotation(label='sigma', start=141, end=147),\n",
       " FunctionAnnotation(label='sigma', start=153, end=157),\n",
       " FunctionAnnotation(label='sigma', start=163, end=201),\n",
       " FunctionAnnotation(label='sigma', start=206, end=263),\n",
       " FunctionAnnotation(label='region', start=236, end=242),\n",
       " FunctionAnnotation(label='region', start=248, end=262),\n",
       " FunctionAnnotation(label='stranded', start=236, end=242),\n",
       " FunctionAnnotation(label='stranded', start=248, end=262),\n",
       " FunctionAnnotation(label='stranded dna', start=236, end=242),\n",
       " FunctionAnnotation(label='stranded dna', start=248, end=262),\n",
       " FunctionAnnotation(label='synthetase', start=168, end=182),\n",
       " FunctionAnnotation(label='synthetase', start=202, end=207),\n",
       " FunctionAnnotation(label='synthetase', start=264, end=271),\n",
       " FunctionAnnotation(label='hydrolysis', start=168, end=179),\n",
       " FunctionAnnotation(label='carboxylic', start=264, end=271)]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keyword_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "69a965bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[FunctionAnnotation(label='trna', start=261, end=265)]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "residue_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3e88684e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1474"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(residue_vocab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9ffcc57f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['active site',\n",
       " 'dimer interface',\n",
       " 'atp binding site',\n",
       " 'metal: magnesium',\n",
       " 'act_site: proton acceptor',\n",
       " 'binding: substrate',\n",
       " 'binding: nad',\n",
       " 'substrate binding site',\n",
       " 'metal binding site',\n",
       " 'homodimer interface',\n",
       " 'np_bind: atp',\n",
       " 'metal: calcium',\n",
       " 'dna binding site',\n",
       " 'np_bind: nad',\n",
       " 'act_site: proton donor',\n",
       " 'mod_res: n6 (pyridoxal phosphate)lysine',\n",
       " 'act_site: nucleophile',\n",
       " 'binding: atp',\n",
       " 'ligand binding site',\n",
       " 'np_bind: nadp',\n",
       " 'abc transporter signature motif',\n",
       " 'walker b',\n",
       " 'd loop',\n",
       " 'np_bind: gtp',\n",
       " 'q loop/lid',\n",
       " 'catalytic residue',\n",
       " 'catalytic site',\n",
       " 'h loop/switch region',\n",
       " 'walker a/p loop',\n",
       " 'phosphorylation site',\n",
       " 'catalytic residues',\n",
       " 'binding: s adenosyl l methionine',\n",
       " 's adenosylmethionine binding site',\n",
       " 'binding: nadp',\n",
       " 'abc atpase subunit interface',\n",
       " 'np_bind: fad',\n",
       " 'binding: fad',\n",
       " 'crosslnk: glycyl lysine isopeptide (lys gly) (interchain with g cter in ubiquitin)',\n",
       " 'binds [4fe 4s] adomet cluster',\n",
       " 'sequence specific dna binding site',\n",
       " 'rna binding site',\n",
       " 'trimer interface',\n",
       " 'walker a motif',\n",
       " 'salt bridge',\n",
       " 'binding: adp; via amide nitrogen',\n",
       " 'walker b motif',\n",
       " 'dimerization interface',\n",
       " 'metal: magnesium; via carbonyl oxygen',\n",
       " 'atp lid',\n",
       " 'g3 box',\n",
       " 'switch ii region',\n",
       " 'non specific dna binding site',\n",
       " 'g2 box',\n",
       " 'binding: adp',\n",
       " 'switch i region',\n",
       " 'conserved gate region',\n",
       " 'dna binding residues',\n",
       " 'metal: zinc; catalytic',\n",
       " 'region: substrate binding',\n",
       " 'g1 box',\n",
       " 'g4 box',\n",
       " 'hexamer interface',\n",
       " 'binding: nad; via amide nitrogen',\n",
       " 'metal: zinc',\n",
       " 'pbp binding loops',\n",
       " 'g5 box',\n",
       " 'tpp binding site',\n",
       " \"pyridoxal 5' phosphate binding site\",\n",
       " 'arginine finger',\n",
       " 'metal: iron sulfur 2 (4fe 4s)',\n",
       " 'binding: adp; via carbonyl oxygen',\n",
       " 'gtp/mg2+ binding site',\n",
       " 'disulfid: redox active',\n",
       " 'nad(p) binding site',\n",
       " 'binding: substrate; via amide nitrogen',\n",
       " 'metal: iron sulfur (4fe 4s s adomet)',\n",
       " 'catalytic triad',\n",
       " 'np_bind: adp',\n",
       " 'coenzyme a binding pocket',\n",
       " 'fes/sam binding site',\n",
       " 'heterodimer interface',\n",
       " 'metal: iron sulfur 3 (4fe 4s)',\n",
       " 'metal: divalent metal cation',\n",
       " 'binding: nad; via carbonyl oxygen',\n",
       " 'heme pocket',\n",
       " 'oligomer interface',\n",
       " 'metal: iron sulfur 1 (4fe 4s)',\n",
       " 'metal: iron (heme axial ligand)',\n",
       " 'region: s adenosyl l methionine binding',\n",
       " 'binding: s adenosyl l methionine; via carbonyl oxygen',\n",
       " 'metal: zinc 2',\n",
       " 'anticodon binding site',\n",
       " 'mg2+ binding site',\n",
       " 'metal: zinc 1; via tele nitrogen',\n",
       " 'binding: gtp; via amide nitrogen',\n",
       " 'metal: manganese 2',\n",
       " 'tetramer interface',\n",
       " 'nucleotide binding site',\n",
       " 'polypeptide substrate binding site',\n",
       " 'ntp binding site',\n",
       " 'metal: zinc 1',\n",
       " 'zn binding site',\n",
       " 'zn2+ binding site',\n",
       " \"act_site: o (3' phospho dna) tyrosine intermediate\",\n",
       " 'chemical substrate binding site',\n",
       " 'binding: heme; covalent',\n",
       " 'act_site: charge relay system',\n",
       " 'pyr/pp interface',\n",
       " 'metal: manganese 1',\n",
       " 'tm abc transporter signature motif',\n",
       " 'motif 2',\n",
       " 'sam binding site',\n",
       " 'metal: iron sulfur 1 (3fe 4s)',\n",
       " 'binding: substrate; via carbonyl oxygen',\n",
       " 'binding: s adenosyl l methionine 2',\n",
       " 'i site',\n",
       " 'substrate binding pocket',\n",
       " 'motif 1',\n",
       " 'metal: zinc 1; via tele nitrogen; catalytic',\n",
       " 'nucleic acid binding site',\n",
       " 'rna binding surface',\n",
       " 'coa binding site',\n",
       " 'motif 3',\n",
       " 'trna binding site',\n",
       " 'metal: manganese',\n",
       " 'gef interaction site',\n",
       " 'binding: fad; via amide nitrogen and carbonyl oxygen',\n",
       " 'metal: magnesium or manganese 1',\n",
       " 'binding: atp; via carbonyl oxygen',\n",
       " 'act_site: acyl ester intermediate',\n",
       " 'metal: magnesium or manganese 2',\n",
       " 'np_bind: nad or nadp',\n",
       " 'mod_res: n methylphenylalanine',\n",
       " 'ligand binding site i',\n",
       " 'nad binding site',\n",
       " 'site: transition state stabilizer',\n",
       " 'fmn binding site',\n",
       " 'catalytic motif',\n",
       " 'binding: fad; via amide nitrogen',\n",
       " 'site: important for catalytic activity',\n",
       " 'binding: s adenosyl l methionine; via amide nitrogen',\n",
       " 'activation loop (a loop)',\n",
       " 'high motif',\n",
       " 'ligand binding site ii',\n",
       " 'act_site: proton donor/acceptor',\n",
       " 'chemical substrate binding pocket',\n",
       " 'np_bind: fmn',\n",
       " 'kmsks motif',\n",
       " 'trna binding surface',\n",
       " 'dxd motif',\n",
       " 'oxyanion hole (oah) forming residues',\n",
       " 'binding: s adenosyl l methionine 2; via carbonyl oxygen',\n",
       " 'protein binding site',\n",
       " 'binding: heme 2; covalent',\n",
       " 'hsp70 interaction site',\n",
       " 'interdomain interaction site',\n",
       " 'metal: magnesium 1',\n",
       " 'binding: heme 1; covalent',\n",
       " 'nadp binding site',\n",
       " 'act_site: for gatase activity',\n",
       " 'binding: s adenosyl l methionine; via amide nitrogen and carbonyl oxygen',\n",
       " 'site: lowers pka of active site tyr',\n",
       " 'mg2+ ligand',\n",
       " 'metal: iron (heme 1 axial ligand)',\n",
       " 'binding: heme c 2; covalent',\n",
       " 'zinc binding site',\n",
       " 'mod_res: 3 oxoalanine (cys)',\n",
       " 'binding: pyridoxal phosphate',\n",
       " '23s rrna interface',\n",
       " 'ssdna binding site',\n",
       " 'binding: glutamine',\n",
       " 'phosphate binding site',\n",
       " 'binding: fmn',\n",
       " 'rrna binding site',\n",
       " \"act_site: o (5' phospho dna) serine intermediate\",\n",
       " 'act_site: phosphoserine intermediate',\n",
       " 'metal: magnesium 2',\n",
       " 'metal: iron (heme 2 axial ligand)',\n",
       " 'iron binding site',\n",
       " 'nudix motif',\n",
       " 'inhibitor cofactor binding pocket',\n",
       " 'lipid binding site',\n",
       " 'binding: atp; via amide nitrogen',\n",
       " 'lipoyl attachment site',\n",
       " 'cytokine receptor motif',\n",
       " 'metal binding triad',\n",
       " 'catalytic nucleophile',\n",
       " 'act_site: cysteine sulfenic acid ( soh) intermediate',\n",
       " 'homotetramer interface',\n",
       " 'metal: iron',\n",
       " 'active site residue',\n",
       " 'mg binding site',\n",
       " 'metal: zinc; via tele nitrogen',\n",
       " 'hpt domain binding site',\n",
       " 'site: essential for catalytic activity',\n",
       " 'generic binding surface i',\n",
       " 'mg2+ ligand, nucleophile: attacks phosphate moiety of intermediate to form covalent intermediate',\n",
       " 'metal: zinc 1; catalytic',\n",
       " 'metal: iron 2',\n",
       " 'active site residues',\n",
       " 'antibiotic binding site',\n",
       " 'metal: divalent metal cation 1',\n",
       " 'metal: iron 1',\n",
       " 'flexible hinge region',\n",
       " 'mod_res: n6 lipoyllysine',\n",
       " 'metal: iron sulfur (4fe 4s)',\n",
       " 'metal: magnesium; via phosphate group',\n",
       " '[2fe 2s] cluster binding site',\n",
       " 'metal: iron sulfur (2fe 2s)',\n",
       " 'acyl acceptor binding pocket',\n",
       " 'mod_res: n6 carboxylysine',\n",
       " 'conserved in all p type atpases',\n",
       " 'coordinates with phosphate oxygen from atp',\n",
       " 'nucleophile: attacks gamma phosphate moiety of atp to form covalent intermediate, mg2+ ligand',\n",
       " 'probable role in phosphate transfer from atp',\n",
       " 'stabilizes transition state of phosphoryl transfer by contacting beta & gamma phosphate oxygens of atp',\n",
       " 'e3 interaction surface',\n",
       " 'metal: divalent metal cation 2',\n",
       " 'helix hairpin helix signature motif',\n",
       " 'metal: calcium 1',\n",
       " 'catalytic loop',\n",
       " 'binding: s adenosyl l methionine 1; via amide nitrogen and carbonyl oxygen',\n",
       " 'metal: zinc 2; catalytic',\n",
       " 'biotinylation site',\n",
       " 'amp binding site',\n",
       " 'binding: s adenosyl l methionine 1',\n",
       " 'binding: heme c; covalent',\n",
       " 'rna/dna hybrid binding site',\n",
       " 'metal: iron (heme axial ligand); via tele nitrogen',\n",
       " 'carboxyltransferase (ct) interaction site',\n",
       " 'lipid: s diacylglycerol cysteine',\n",
       " 'giy yig motif/motif a',\n",
       " 'binding: substrate; via amide nitrogen and carbonyl oxygen',\n",
       " 'metal: molybdenum; via amide nitrogen',\n",
       " 'tetramer (dimer of dimers) interface',\n",
       " '16s rrna binding site',\n",
       " 'metal: potassium ion; via carbonyl oxygen',\n",
       " 'domain linker motif',\n",
       " 'minor groove reading motif',\n",
       " 'binding: atp; via amide nitrogen and carbonyl oxygen',\n",
       " 'active site pocket',\n",
       " 'phosphopeptide binding site',\n",
       " 'mod_res: phosphohistidine; by autocatalysis',\n",
       " '5s rrna interface',\n",
       " 'generic binding surface ii',\n",
       " 'binding: heme c 1; covalent',\n",
       " 'nad binding pocket',\n",
       " 'oxyanion hole',\n",
       " 'sugar binding site',\n",
       " 'n terminal domain interface',\n",
       " 'lipid: n palmitoyl cysteine',\n",
       " 'fe s cluster binding site',\n",
       " 'trd cr/trd cr interface',\n",
       " 'metal: iron sulfur 1 (4fe 4s s adomet)',\n",
       " 'binding: camp 2',\n",
       " 'c terminal domain interface',\n",
       " 'metal: molybdenum',\n",
       " 'homotrimer interaction site',\n",
       " 'interdomain contacts',\n",
       " 'metal ion dependent adhesion site (midas)',\n",
       " 'peptide binding site',\n",
       " 'metal: potassium; via carbonyl oxygen',\n",
       " 'binds [4fe 4s] cluster',\n",
       " 'domain interface',\n",
       " 'metal: zinc; via tele nitrogen; catalytic',\n",
       " 'regulatory phosphorylation site',\n",
       " 'mg++ binding site',\n",
       " 'cosubstrate binding site',\n",
       " \"pyridoxal 5' phosphate (plp) binding site\",\n",
       " 'metal: molybdenum; via carbonyl oxygen',\n",
       " 'act_site: tele phosphohistidine intermediate',\n",
       " 'catalytic tetrad',\n",
       " 'substrate cofactor binding pocket',\n",
       " 'metal binding ligand',\n",
       " 'region: s adenosyl l methionine 2 binding',\n",
       " 'metal: cobalt (cobalamin axial ligand)',\n",
       " 'gsh binding site (g site)',\n",
       " 'region: pyridoxal phosphate binding',\n",
       " 'metal: iron sulfur 2',\n",
       " 'binding: cobalamin',\n",
       " 'metal: iron sulfur 1',\n",
       " 'ca2+ binding site',\n",
       " 'metal: copper',\n",
       " 'metal: molybdate',\n",
       " 'act_site: acyl thioester intermediate',\n",
       " 'gamma subunit interface',\n",
       " 's8 interaction site',\n",
       " 'intersubunit interface',\n",
       " 'metal: metal cation',\n",
       " 'synthetase active site',\n",
       " 'anchoring element',\n",
       " 'catalytic dyad',\n",
       " 'shikimate binding site',\n",
       " 'act_site: for beta ketoacyl synthase activity',\n",
       " 'phosphate binding motif',\n",
       " 'acyl activating enzyme (aae) consensus motif',\n",
       " 'cohesin binding interface 1',\n",
       " 'mpt binding site',\n",
       " 'beta alpha beta structure motif',\n",
       " 'act_site: thioimidate intermediate',\n",
       " 'helical arch',\n",
       " 'motif: nudix box',\n",
       " 'nucleic acid substrate binding site',\n",
       " 'catalytic core',\n",
       " 'fad binding site',\n",
       " 'coenzymea binding site',\n",
       " 'site: activates thiol group during catalysis',\n",
       " 'alpha subunit interface',\n",
       " 'phb binding site',\n",
       " 'nucleotidyl binding site',\n",
       " 'l27 interface',\n",
       " 'metal: magnesium or manganese',\n",
       " 'aspartate rich region 2',\n",
       " 'fad binding pocket',\n",
       " 'l25 interface',\n",
       " 'metal: zinc 2; via tele nitrogen',\n",
       " 'binding: pyrroloquinoline quinone',\n",
       " 'active sites',\n",
       " 'site: important for beta aspartyl amp intermediate formation',\n",
       " 'aspartate rich region 1',\n",
       " 'subunit interaction site',\n",
       " 'cohesin binding interface 2',\n",
       " 'fad binding motif',\n",
       " 'heme binding site',\n",
       " 'binding: fmn; shared with dimeric partner',\n",
       " 'nucleotide binding pocket',\n",
       " 'metal: iron sulfur (3fe 4s)',\n",
       " 'metal: iron (heme 2 axial ligand); via tele nitrogen',\n",
       " 'active site lid residues',\n",
       " 'site: important for catalysis',\n",
       " 'binding surface',\n",
       " 'adp binding site',\n",
       " 'domain iii',\n",
       " 'substrate mg2+ binding site',\n",
       " 'metal: calcium 2',\n",
       " 'molecular tunnel',\n",
       " 'mobile loop',\n",
       " 'adp binding pocket',\n",
       " 'site: participates in a stacking interaction with the thymidine ring of dtdp 4 oxo 6 deoxyglucose',\n",
       " 'presynaptic site i dimer interface',\n",
       " 'metal: iron sulfur 2 (2fe 2s)',\n",
       " 'domain ii',\n",
       " '23s rrna binding site',\n",
       " 'synaptic antiparallel dimer interface',\n",
       " 'synaptic flat tetramer interface',\n",
       " 'probable catalytic site',\n",
       " 'laminin 1 binding site',\n",
       " 'synaptic site i dimer interface',\n",
       " '(t/h)xgh motif',\n",
       " 'region: cobalamin binding',\n",
       " 'conserved cys residue',\n",
       " 'ef tu interaction site',\n",
       " 'ef g interaction site',\n",
       " 'l11 interface',\n",
       " 'binding: glyceraldehyde 3 phosphate',\n",
       " 'multimer interface',\n",
       " 'binding: nadp; via amide nitrogen and carbonyl oxygen',\n",
       " 'nucleoside/zn binding site',\n",
       " 'chain length determination region',\n",
       " 'dnaa box binding interface',\n",
       " 'site: raises pka of active site his',\n",
       " 'folate binding site',\n",
       " 'domain iv',\n",
       " 'metal: calcium 1; via carbonyl oxygen',\n",
       " 'roof hairpin',\n",
       " 'metal: zinc; via pros nitrogen',\n",
       " 'mg2+ ligand, nucleophile: attacks phosphate moiety of substrate to form covalent intermediate',\n",
       " 'pbp binding regions',\n",
       " 'hydrophobic ligand binding site',\n",
       " 'sula interaction site',\n",
       " 'gtp binding site',\n",
       " 'hsp70 (atpase domain) interactions',\n",
       " 'domain i',\n",
       " 'rna binding motif',\n",
       " 'dna binding helix',\n",
       " 'act_site: 5 glutamyl coenzyme a thioester intermediate',\n",
       " 'epsilon subunit interface',\n",
       " 'b12 binding site',\n",
       " 'zn_fing: chc2 type',\n",
       " 'cap like domain',\n",
       " 'cobalt ligand',\n",
       " 'deacylase active site',\n",
       " 'region: glyceraldehyde 3 phosphate binding',\n",
       " 'site: important for catalytic activity, responsible for pka modulation of the active site glu and correct orientation of both the proton donor and substrate',\n",
       " 'hinge regions',\n",
       " 'act_site: phosphocysteine intermediate',\n",
       " 'dna binding groove',\n",
       " 'the sites determining sugar specificity',\n",
       " 'l5 interface',\n",
       " 'binding: heme 3; covalent',\n",
       " 'inhibitor binding site',\n",
       " 's12 interface',\n",
       " 'anti sigma factor interaction site',\n",
       " 'substrate binding region',\n",
       " '3fe 4s cluster binding site',\n",
       " 'a component interface',\n",
       " 'metal binding site 2',\n",
       " 'binding: heme (covalent)',\n",
       " 'magnesium binding site',\n",
       " 'na binding site',\n",
       " 'hinge',\n",
       " 'primary dimer interface',\n",
       " 'tetrahydrofolate binding site',\n",
       " 'metal ligand',\n",
       " 'l10 interface',\n",
       " 'hinge region',\n",
       " 'abstracts alpha proton (base)',\n",
       " 'act_site: gmp histidine intermediate',\n",
       " 'rrna interaction site',\n",
       " 'mpn+ (jamm) motif',\n",
       " 'adp ribose binding site',\n",
       " 'metal: iron (heme 3 axial ligand); via tele nitrogen',\n",
       " 'metal binding residues',\n",
       " 'cmp binding site',\n",
       " 'sbd interface',\n",
       " 'atp amp (ap5a) binding site',\n",
       " 'core dimer interface',\n",
       " 'glutaminase active site',\n",
       " 'homohexameric interface',\n",
       " 'metal: iron sulfur (2fe 2s); shared with dimeric partner',\n",
       " 'php thumb interface',\n",
       " 'site: essential for prephenate dehydratase activity',\n",
       " 'metal: divalent metal cation; via tele nitrogen',\n",
       " 'l16 interface',\n",
       " 'l3 interface',\n",
       " 'g x2 g x g k',\n",
       " 'peripheral dimer interface',\n",
       " 'oligomerisation interface',\n",
       " 'metal binding site 1',\n",
       " 'general acid/base',\n",
       " 'rnase e interface',\n",
       " 'substrate interface',\n",
       " 'zn binding residues',\n",
       " \"a' component interface\",\n",
       " 'rpb12 interaction site',\n",
       " 'lbp interface',\n",
       " 'metal: magnesium 2; catalytic',\n",
       " 'amidase catalytic site',\n",
       " 'chew interface',\n",
       " 'neub binding interface',\n",
       " 'metal: magnesium 1; catalytic',\n",
       " 'binding: nad 1',\n",
       " 'ctc domain interface',\n",
       " 'active site turn',\n",
       " 'molybdopterin cofactor binding site',\n",
       " 'oligonucleotide binding site',\n",
       " 'binding: nadp; via amide nitrogen',\n",
       " 'substrate binding pocket (h site)',\n",
       " 'binding: heme 4; covalent',\n",
       " 'tmp binding site',\n",
       " 'pyrophosphate binding site',\n",
       " 'tandem repeat interface',\n",
       " 'atp binding subdomain',\n",
       " 'candidate oxyanion hole',\n",
       " 'nodb motif',\n",
       " 'n and c terminal domain interface',\n",
       " 'p loop',\n",
       " 'interface with l7/l12 ribosomal proteins',\n",
       " 'thioester formation/cholesterol transfer',\n",
       " 'dimerization interface 35a',\n",
       " 'n terminal plug',\n",
       " 'conserved cis peptide bond',\n",
       " 'atp binding pocket',\n",
       " 'rpb3 interaction site',\n",
       " 'active site loop',\n",
       " 'metal: molybdate; via amide nitrogen',\n",
       " 'l13 binding site',\n",
       " 'potential glutamine specificity residues',\n",
       " 'beta subunit interface',\n",
       " 'catalytic cysteine',\n",
       " 'smpb tmrna interface',\n",
       " 'g loop',\n",
       " 'stacking interactions',\n",
       " 'k x d g motif',\n",
       " 'cdp binding sites',\n",
       " 'homohexamer interface',\n",
       " 'coa binding pocket',\n",
       " 'metal: divalent metal cation; via carbamate group',\n",
       " 'np_bind: gmp',\n",
       " 'cu binding site',\n",
       " 'dimerization subdomain',\n",
       " '16s/18s rrna binding site',\n",
       " 'l phe binding site',\n",
       " 'rpb10 interaction site',\n",
       " 'ligand binding sites',\n",
       " 'hydrophobic core',\n",
       " 'dead box helicase motif',\n",
       " 'l7/l12 interface',\n",
       " 'metal: iron (heme 4 axial ligand); via tele nitrogen',\n",
       " 'rxxxh motif',\n",
       " \"alphantd beta' interaction site\",\n",
       " 'glutamase interaction surface',\n",
       " 'act_site: phosphothreonine intermediate',\n",
       " 'intramolecular domain interface',\n",
       " 'rpb11 interaction site',\n",
       " 'ca binding site',\n",
       " 'thiostrepton binding site',\n",
       " 'phosphorylation loop region',\n",
       " 'amp/ppi binding site',\n",
       " 'diiron binding motif',\n",
       " 'homotetrameric interface',\n",
       " 'act_site: schiff base intermediate with acetaldehyde',\n",
       " 'active site lid',\n",
       " 's component interface',\n",
       " 'atp binding',\n",
       " 'imp binding site',\n",
       " 'site: lowers pka of active site cys',\n",
       " 's18 interface',\n",
       " 'carbohydrate binding site',\n",
       " 'binding: dihydroxyacetone phosphate; via amide nitrogen',\n",
       " 'region: dihydroxyacetone phosphate binding',\n",
       " 'na+ binding site',\n",
       " 'substrate binding site 2',\n",
       " 'l21 binding site',\n",
       " 'act_site: o isoaspartyl threonine intermediate',\n",
       " 'mod_res: o amp tyrosine',\n",
       " 'act_site: schiff base intermediate with d ribose 5 phosphate',\n",
       " 'catalytic center binding site',\n",
       " 'act_site: pros phosphohistidine intermediate',\n",
       " 'pentamer interface',\n",
       " 'binding: gtp; via carbonyl oxygen',\n",
       " 'metal: iron (heme axial ligand); shared with second transmembrane subunit',\n",
       " 'catalytic zn binding site',\n",
       " 'binding: coenzyme a; via amide nitrogen',\n",
       " 'homopentamer interface',\n",
       " 'mod_res: fmn phosphoryl threonine',\n",
       " 'translocon binding site',\n",
       " 'dna binding surface',\n",
       " 'alphantd beta interaction site',\n",
       " 's9 interface',\n",
       " 'binding: gtp',\n",
       " 'metal: iron (heme 1 axial ligand); via tele nitrogen',\n",
       " 'binding: alpha ketoisovalerate',\n",
       " 'region: alpha ketoisovalerate binding',\n",
       " 'binding: spermidine',\n",
       " 'metal: iron 1 (heme b562 axial ligand)',\n",
       " 'partial ornithine binding site',\n",
       " 'substrate binding site 1',\n",
       " 'mrna/rrna interface',\n",
       " 'structural zn2+ binding site',\n",
       " 'hydrogen bonds to substrate diphosphate',\n",
       " 'metal ligand (via water)',\n",
       " 'metal ligand; stabilizes partially closed conformation via electrostatic interaction with k330',\n",
       " 'stabilizes fully closed conformation via electrostatic interaction with c terminal carboxylate of ipp',\n",
       " 'stabilizes partially closed conformation via electrostatic interactions with k330',\n",
       " 'binding: cobalamin; via carbonyl oxygen',\n",
       " 'trigger factor interaction site',\n",
       " 'cofactor binding site',\n",
       " 's13e l30e interaction site',\n",
       " 'site: important for autoinhibition of adenylyltransferase activity',\n",
       " 'metal: metal cation; via carbonyl oxygen',\n",
       " 'metal: iron 2 (heme b566 axial ligand)',\n",
       " 'l23 interface',\n",
       " 'metal: cobalt',\n",
       " 'regulatory protein interface',\n",
       " 'central domain interface',\n",
       " 'effector binding site',\n",
       " 'l serine binding site',\n",
       " 'binding: fmn; via amide nitrogen and carbonyl oxygen',\n",
       " 'glutamine binding',\n",
       " 'purine monophosphate binding site',\n",
       " 'valine binding site',\n",
       " 'act_site: cysteine persulfide intermediate',\n",
       " 'cleavage site',\n",
       " 'mg2+/mn2+ binding site',\n",
       " 'signal recognition particle (srp54) interaction site',\n",
       " 'metal: zinc 2; via tele nitrogen; catalytic',\n",
       " 'beta clamp/translesion dna polymerase binding surface',\n",
       " 'uridine monophosphate binding site',\n",
       " 'oxyanion strand',\n",
       " 'trimerization site',\n",
       " \"rpb1 (beta') rpb2 (beta) interaction site\",\n",
       " 'protein rrna interface',\n",
       " 'iron coordination center',\n",
       " 'rpb1 interaction site',\n",
       " 'metal: zinc; via tele nitrogen; shared with neighboring monomer',\n",
       " 'oligomerization interface',\n",
       " 'beta clamp/clamp loader binding surface',\n",
       " 'protein interface 1',\n",
       " 's11 interface',\n",
       " 'allosteric regulatory site',\n",
       " 'site: contributes to substrate recognition',\n",
       " 'site: stabilizes the phosphoryl group',\n",
       " 'subunit interface',\n",
       " 'act_site: schiff base intermediate with substrate',\n",
       " 'mod_res: o ump tyrosine',\n",
       " 'glycine pyridoxal phosphate binding site',\n",
       " 'udp glcnac binding site',\n",
       " 'product (indole) binding pocket',\n",
       " 'metal: iron 3',\n",
       " '30s subunit binding site',\n",
       " 'mgatp binding site',\n",
       " 'inhibitor site',\n",
       " 'atp/mg binding site',\n",
       " 'nef interaction site',\n",
       " 'cu(i) binding site',\n",
       " 'translocon interaction site',\n",
       " 'functional site',\n",
       " 'binding: camp 1',\n",
       " 'thiamine phosphate binding site',\n",
       " 'binding: ubiquinone',\n",
       " 'nadp+ binding site',\n",
       " 'effector binding pocket',\n",
       " 'rodz interaction site',\n",
       " 'zn binding sites',\n",
       " 'structural zn binding site',\n",
       " 'metal: zinc 2; via pros nitrogen',\n",
       " '25s rrna binding site',\n",
       " 'metal: nickel iron sulfur (ni 4fe 4s)',\n",
       " 'product binding site',\n",
       " 'h2th interface',\n",
       " 'binding: heme',\n",
       " 'nucleophile elbow',\n",
       " 'hit family signature motif',\n",
       " 'crosslnk: isoglutamyl lysine isopeptide (lys gln) (interchain with q cter in protein pup)',\n",
       " 'binding: pep',\n",
       " 'region: pep binding',\n",
       " 'site: transition state stablizer',\n",
       " 'mg2+ ligand, general acid: donates proton to leaving group, general base: activates water',\n",
       " 'aspartate binding site',\n",
       " 'metal: magnesium 2; via carbonyl oxygen; catalytic',\n",
       " 'ring oligomerisation interface',\n",
       " 'substrate (anthranilate) binding pocket',\n",
       " 'turnover facilitating residue',\n",
       " '8og recognition residue',\n",
       " 'binding: coenzyme a',\n",
       " 'ribulose/triose binding site',\n",
       " \"pyridoxal 5' phosphate binding pocket\",\n",
       " 'gsh binding site',\n",
       " 'proximal heme binding site',\n",
       " 'acid (promotes elimination of oh)',\n",
       " 'metal binding ligand (2nd metal)',\n",
       " 'site: important for activity',\n",
       " 'act_site: amp threonine intermediate',\n",
       " 'active site flap',\n",
       " 'delta subunit interface',\n",
       " 'na binding site 1',\n",
       " 'metal: calcium; via carbonyl oxygen',\n",
       " 'substrate binding/catalytic site',\n",
       " 'mod_res: 2,3 didehydroalanine (ser)',\n",
       " 'site: critical for catalysis',\n",
       " 'aminoacyl trna interaction site (a site)',\n",
       " 'metal: iron (heme proximal ligand)',\n",
       " 'iron sulfur protein interface',\n",
       " 'protofilament interaction site',\n",
       " 'signature motif',\n",
       " 'allosteric binding site',\n",
       " 'binuclear metal center',\n",
       " 'act_site: proton acceptor; specific for (s) substrate epimerization',\n",
       " 'act_site: proton acceptor; specific for (r) substrate epimerization',\n",
       " 'alphantd homodimer interface',\n",
       " '23s rrna interaction site',\n",
       " 'streptomycin interaction site',\n",
       " 'homotrimer interface',\n",
       " 'ap binding site',\n",
       " 'pore gating glutamate residue',\n",
       " 'tyrosyl radical',\n",
       " \"beta' and sigma factor interface\",\n",
       " 'allosteric site',\n",
       " 'intercalation triad',\n",
       " 'cl binding site',\n",
       " 'crosslnk: 5 imidazolinone (ala gly)',\n",
       " 'reading head residues',\n",
       " \"beta and beta' interface\",\n",
       " 'active site region',\n",
       " 'binding: imp',\n",
       " 'region: imp binding',\n",
       " 'mod_res: 3 oxoalanine (ser)',\n",
       " 'act_site: schiff base intermediate with substrate; via pyruvic acid',\n",
       " 'nucleophile, forms methylthioether with methyl group from sam',\n",
       " 'nucleophile, initiates reductive cleavage of s s bond',\n",
       " 'parb box ii like motif',\n",
       " 'homodimer binding site',\n",
       " 'site: catalytically relevant',\n",
       " 'metal binding site b',\n",
       " '3 fold/trimer interface',\n",
       " 'act_site: proton donor; for transglycosylase activity',\n",
       " 'act_site: acyl ester intermediate; for transpeptidase activity',\n",
       " 'domain interaction interfaces',\n",
       " 'na binding site 2',\n",
       " 'heterotetramer interface',\n",
       " 'oligomeric interface',\n",
       " '4 fold oligomerization interface',\n",
       " \"mod_res: o (pantetheine 4' phosphoryl)serine\",\n",
       " '16s rrna interaction site',\n",
       " 'purine nucleoside binding site',\n",
       " 'kmsk motif region',\n",
       " 'binding: gmp',\n",
       " 'gtpase/zn binding domain interface',\n",
       " 'tetramerization interface',\n",
       " 'iron sulfur cluster',\n",
       " 'mod_res: pyruvic acid (ser)',\n",
       " 'positions lys',\n",
       " 'general acid, sometimes general base, mg2+ ligand',\n",
       " 'chain: aspartate 1 decarboxylase beta chain',\n",
       " 'diiron center',\n",
       " 'cpxp motif',\n",
       " 'dna binding and oxog recognition site',\n",
       " 'nucleotide binding site/active site',\n",
       " 'hsdm interface',\n",
       " 'chain: aspartate 1 decarboxylase alpha chain',\n",
       " 'core domain interface',\n",
       " 'coenzyme a binding site',\n",
       " 'binding: fad; via carbonyl oxygen',\n",
       " 'site: important for substrate specificity',\n",
       " 'cation binding site',\n",
       " 'hexamer (dimer of trimers) interface',\n",
       " 'substrate interaction site',\n",
       " 'structural tetrad',\n",
       " 'ugi interface',\n",
       " 'binding: l threonine',\n",
       " 'binding: l threonine; via carbonyl oxygen',\n",
       " 'interface (dimer of trimers)',\n",
       " 'sulfate 1 binding site',\n",
       " 's17 interaction site',\n",
       " 'di iron ligands',\n",
       " 'hxxh motif',\n",
       " 'metal: calcium 2; via carbonyl oxygen',\n",
       " 'dinuclear metal binding motif',\n",
       " 'act_site: nucleophile and sulfur donor',\n",
       " 'region: pyrroloquinoline quinone binding',\n",
       " 'homodimer homodimer interface',\n",
       " 'metal binding site a',\n",
       " 'this thif interaction site',\n",
       " 'activator binding site',\n",
       " 'binding: coenzyme a; via carbonyl oxygen',\n",
       " 'site: cleavage; by autolysis',\n",
       " 'metal: iron (heme 3 axial ligand)',\n",
       " 'radical transfer pathway',\n",
       " 'peroxidatic and resolving cysteines',\n",
       " 'binding: l tryptophan',\n",
       " 'fe binding site',\n",
       " 'allosteric switch controlling residues',\n",
       " 'predicted 30s ribosome binding site',\n",
       " 'dimerization domain swap beta strand',\n",
       " 'helps position phosphoryl group',\n",
       " 'camp binding site',\n",
       " 'n acetyl l glutamate binding site',\n",
       " 'proximal quinone binding site',\n",
       " 'dcoh dimer interaction site',\n",
       " '8 oxo dgmp binding site',\n",
       " 'acid (promotes elim of oh)',\n",
       " 'facilitates removal of mg2+ ppi product complex formed in both half reactions',\n",
       " 'dcoh /hnf 1 dimer interaction site',\n",
       " 'decamer (pentamer of dimers) interface',\n",
       " 'dcoh tetramer interaction site',\n",
       " 'controls pka of catalytic his',\n",
       " 'active site clefts',\n",
       " 'non specific dna interactions',\n",
       " 'np_bind: fmn binding',\n",
       " 'aromatic arch',\n",
       " 'binding: heme 2 (covalent)',\n",
       " 'binds phosphate moiety of substrate, positions water',\n",
       " 'binding: heme 1 (covalent)',\n",
       " 'ferroxidase diiron center',\n",
       " 'substrate binding site i',\n",
       " 'positions nucleophilic asp, binds phosphate moiety of substrate',\n",
       " 'mg ion binding site',\n",
       " 'gxxg motif',\n",
       " 'substrate binding site ii',\n",
       " 'dockerin binding interface',\n",
       " 'binding: d ribulose 5 phosphate',\n",
       " 'region: d ribulose 5 phosphate binding',\n",
       " 'domain 2 interface',\n",
       " 'moad interaction',\n",
       " 'hpr binding site',\n",
       " 'this thig interaction site',\n",
       " 'metal: zinc 4',\n",
       " 'domain crossover interface',\n",
       " 'binuclear metal binding site',\n",
       " 'switch regulator',\n",
       " 'proton abstraction (general base)',\n",
       " 'tetramer interfaces',\n",
       " 'moae homodimer interface',\n",
       " 'l21e interface',\n",
       " 'pantoate binding site',\n",
       " 'site: may be important for catalysis',\n",
       " 'glycerol binding site',\n",
       " 'homohexamer subunit interaction site',\n",
       " 'site: important for substrate recognition',\n",
       " 'binding pocket',\n",
       " 'nucleophile',\n",
       " 'quinone interaction residues',\n",
       " \"rpb1 (beta') rpb6 (omega) interaction site\",\n",
       " 'hexagonal pore residue',\n",
       " 'binding: heme (covalent; via pros nitrogen)',\n",
       " 'metal: nickel; via tele nitrogen',\n",
       " 'anp binding site',\n",
       " 'fbp binding site',\n",
       " 'act_site: amidino cysteine intermediate',\n",
       " 'site: critical for activity',\n",
       " 'subunit i/iii interface',\n",
       " 'citrylcoa binding site',\n",
       " 'homodimer interaction site',\n",
       " 'binding: allantoate',\n",
       " 'binding: allantoate; via carbonyl oxygen',\n",
       " 'act_site: proton acceptor; via imino nitrogen',\n",
       " 'moad moeb interaction site',\n",
       " 'metal binding active site',\n",
       " 'moad moae interaction site',\n",
       " 'domain 3 interface',\n",
       " 'cataytic base',\n",
       " 'dna intercalation site',\n",
       " 'interacts with phosphate oxygen, facilitating hydrolysis',\n",
       " 'metal: iron (heme 5 axial ligand)',\n",
       " 'act_site: schiff base intermediate with dna',\n",
       " 'trimer interface a',\n",
       " 'dps ferroxidase diiron center',\n",
       " 'act domain interface',\n",
       " 'trimer interface b',\n",
       " 'domain 1 interface',\n",
       " 'sm1 motif',\n",
       " 'binding: heme 5; covalent',\n",
       " 'heme binding residues',\n",
       " 'cl binding residues',\n",
       " 'cl selectivity filter',\n",
       " 'nadh binding site',\n",
       " 'sm2 motif',\n",
       " 'dna interaction',\n",
       " 'binds s adensosylmethionine',\n",
       " 'act_site: n6 amp lysine intermediate',\n",
       " 'mismatch recognition site',\n",
       " 'thf binding site',\n",
       " 'trnatyr binding site',\n",
       " 'schiff base residues',\n",
       " 'redox dependent activation switch',\n",
       " 'thiamine binding site',\n",
       " 'hm ion binding site',\n",
       " 'metal: divalent metal cation 1; via tele nitrogen',\n",
       " 'metal: divalent metal cation 2; via tele nitrogen',\n",
       " 'metal: divalent metal cation 2; via pros nitrogen',\n",
       " 'transmembrane helices',\n",
       " 'eyelet of channel',\n",
       " 'metal: iron (heme axial ligand); shared with dimeric partner',\n",
       " 'feedback inhibition sensing region',\n",
       " 'region: chorismate binding',\n",
       " 'region: tryptophan binding',\n",
       " 'metal: zinc 1; via carbamate group',\n",
       " 'metal: zinc 2; via carbamate group',\n",
       " 'q x w motif',\n",
       " 'intersubunit interactions',\n",
       " 'this interaction site',\n",
       " 'protein iiaglc interface',\n",
       " 'hydrophilic channel',\n",
       " 'sugar binding sites',\n",
       " 'binding: l tryptophan; via carbonyl oxygen',\n",
       " 'ion pore',\n",
       " 'forms cationic cavity that facilitates hydrolysis of phosphate',\n",
       " 'active site nucleophile',\n",
       " 'effector interaction site',\n",
       " 'nuceophile',\n",
       " 'nucleophile, methyl donor',\n",
       " 'oxalacetate/citrate binding site',\n",
       " 'cysd dimerization site',\n",
       " 'metal binding sites',\n",
       " 'calcium binding site',\n",
       " 'metal: nickel',\n",
       " 'additional dna contacts',\n",
       " 'substrate specificity',\n",
       " 'rab family motif 5 (rabf5)',\n",
       " 'rab family motif 4 (rabf4)',\n",
       " 'qo binding site',\n",
       " 'glycine loop',\n",
       " 'base (abstracts alpha proton), acid (donates proton to leaving group)',\n",
       " 'stabilizes intermediate through cation pi interaction',\n",
       " 'pyrimidine base specificity',\n",
       " 'dimer interaction site',\n",
       " 'uracil binding',\n",
       " 'steroid binding site',\n",
       " 'flim/flig interface',\n",
       " 'act_site: glycyl thioester intermediate',\n",
       " 'metal: iron (heme 4 axial ligand)',\n",
       " 'catalytic base',\n",
       " 'rab family motif 3 (rabf3)',\n",
       " 'threonine allosteric regulatory site',\n",
       " 'nef/hsp70 interaction site',\n",
       " 'rab family motif 2 (rabf2)',\n",
       " 'hydrophobic gate',\n",
       " 'proposed active site lysine',\n",
       " 'xylulose binding site',\n",
       " 'primase nucleotide binding site',\n",
       " 'ligand binding surface',\n",
       " 'proposed catalytic triad',\n",
       " 'periplasmic entrance',\n",
       " 'beta sandwich domain interface',\n",
       " 'metal: iron (heme c 1 axial ligand); via tele nitrogen',\n",
       " 'gdi interaction site',\n",
       " 'ligand binding cavity',\n",
       " 'substrate binding tunnel',\n",
       " 'sugar specificity',\n",
       " 'tetrameric interface',\n",
       " 'sdhd (cybs) interface',\n",
       " 'switch loop',\n",
       " 'dsbd alpha interface',\n",
       " 'intrachain domain interface',\n",
       " 'rab subfamily motif 3 (rabsf3)',\n",
       " 'rab family motif 1 (rabf1)',\n",
       " 'protein splicing catalytic site',\n",
       " '[4fe 4s] binding site',\n",
       " 'metal: iron (heme c 2 axial ligand); via tele nitrogen',\n",
       " 'metal: iron sulfur 4 (4fe 4s)',\n",
       " 'mod_res: adp ribosylarginine; by dinitrogenase reductase adp ribosyltransferase',\n",
       " 'binds zn(ii) ion',\n",
       " 'act_site: 4 aspartylphosphate intermediate',\n",
       " 'allosteric regulatory residue',\n",
       " 'nad+ binding site',\n",
       " 'rab subfamily motif 2 (rabsf2)',\n",
       " 'interchain domain interface',\n",
       " 'nucleophilic elbow',\n",
       " 'region: phosphate binding',\n",
       " 'binding: phosphate',\n",
       " 'r2 peptide binding site',\n",
       " 'region: l glutamine binding',\n",
       " 'binding: l glutamine',\n",
       " 'pa/protease or protease like domain interface',\n",
       " 'activates water (indirectly)',\n",
       " 'active pocket/dimerization site',\n",
       " 'thioester region',\n",
       " 'polymerase nucleotide binding site',\n",
       " 'binding: substrate 1',\n",
       " 'binding: substrate 2',\n",
       " 'trna',\n",
       " 'hslu subunit interaction site',\n",
       " 'basic sphincter',\n",
       " 'site: important for enzyme activity',\n",
       " 'metal: iron (heme c 1 axial ligand)',\n",
       " 'metal: iron (heme c 2 axial ligand)',\n",
       " 'ribose interaction site',\n",
       " 'binding: phosphatidylinositol 4,5 bisphosphate',\n",
       " 'active site 1',\n",
       " 'homooctamer interface',\n",
       " 'rna binding cleft',\n",
       " 'binding: substrate; via tele nitrogen',\n",
       " 'interacts with substrate/intermediate',\n",
       " 'nucleophile: forms covalent intermediate',\n",
       " 'distal heme binding site',\n",
       " 'metal: zinc 2; via pros nitrogen; catalytic',\n",
       " 'ap4a binding site',\n",
       " 'radical stabilisation',\n",
       " 'acetate binding site',\n",
       " 'd x s d e motif',\n",
       " 'triphosphate binding site',\n",
       " 'heme bl binding site',\n",
       " 'subunit iii/vb interface',\n",
       " 'site: plays an important role in maintaining the position of the catalytic nucleophile',\n",
       " 'acceptor phosphate pocket',\n",
       " 'active site cysteine',\n",
       " 'ligand binding pocket/active site',\n",
       " 'polyp binding site',\n",
       " 'hexagonal pore',\n",
       " 'ap (apurinic/apyrimidinic) site pocket',\n",
       " 'probable active site',\n",
       " 'pua domain interface',\n",
       " 'ile/val binding site',\n",
       " 'dna binding interface',\n",
       " 'skp1 binding site',\n",
       " 'mod_res: o (phosphoribosyl dephospho coenzyme a)serine',\n",
       " 'np_bind: nad 1',\n",
       " 'subunit i/ii interface',\n",
       " 'binding: purine nucleoside',\n",
       " 'binding: phosphate; via amide nitrogen',\n",
       " 'coenzyme q binding site',\n",
       " 'surface patch',\n",
       " 'nickel binding site',\n",
       " 'site: enhances dephosphorylation of chey p',\n",
       " 'plp binding',\n",
       " 'plp binding, reaction nucleophile',\n",
       " 'binding: d ribose 5 phosphate',\n",
       " 'binding: d ribose 5 phosphate; via amide nitrogen',\n",
       " 'region: d ribose 5 phosphate binding',\n",
       " 'heptamer interface',\n",
       " 'water exit pathway',\n",
       " 'subunit iii/vib interface',\n",
       " 'binding: substrate; shared with homodimeric partner',\n",
       " 'active site 2',\n",
       " 'mn binding site',\n",
       " 'active site tetrad',\n",
       " 'qi binding site',\n",
       " 'proton exit pathway',\n",
       " 'metal: nickel iron sulfur (ni 4fe 4s); via tele nitrogen',\n",
       " 'heterodimer interface 2',\n",
       " 'subunit iii/via interface',\n",
       " 'tdp binding site',\n",
       " 'mod_res: 4 aspartylphosphate',\n",
       " 'mod_res: phosphohistidine',\n",
       " 'act_site: n6 gmp lysine intermediate',\n",
       " 'active site flap/lid',\n",
       " 'ferroxidase pore',\n",
       " 'hydroxycobalamin binding site',\n",
       " 'heterodimer interface 1',\n",
       " 'metal: iron 2; via pros nitrogen',\n",
       " 'metal: calcium 3',\n",
       " 'aspartate rich active site metal binding site',\n",
       " 'dimerization domain',\n",
       " 'gdp fucose binding site',\n",
       " 'multimerization interface',\n",
       " 'heterotrimer interface',\n",
       " 'alpha helical domain',\n",
       " 'pyrimidine nucleoside binding site',\n",
       " 'had signature motif i',\n",
       " ...]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "residue_vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "975f93d2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esm3_peft",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
